zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/Leader.java

```diff
@@ -614,7 +614,7 @@ void lead() throws IOException, InterruptedException {
 
             QuorumVerifier lastSeenQV = self.getLastSeenQuorumVerifier();
             QuorumVerifier curQV = self.getQuorumVerifier();
-            if (curQV.getVersion() == 0 && curQV.getVersion() == lastSeenQV.getVersion()) {
+            if (!QuorumPeerConfig.isReconfigEnabled() || (curQV.getVersion() == 0 && curQV.getVersion() == lastSeenQV.getVersion())) {
```

Comment:
Could you explain the "!QuorumPeerConfig.isReconfigEnabled() || " here ? And not, e.g., "QuorumPeerConfig.isReconfigEnabled() &&"

The previous condition means this is done only when the config is an initial one. What's being done here is updating the config version to the current leader zxid. This is done only once, when the ensemble is initial.

Reply:
Thanks for checking!

> What's being done here is updating the config version to the current leader zxid

I added the extra condition because here also the lastSeenQuorumVerifier is set to be equal to the currentQuorumVerifyer. The problem I found in ZOOKEEPER-3814 was that old server addresses were still propagated through the lastSeenQuorumVerifier to newly restarted nodes when dynamic reconfig was disabled. So my idea was to reset the lastSeenQuorumVerifier to the current config here. Maybe a better way would be to not update lastSeenQuorumVerifier when we process the NEWLEADER message (if dynamic reconfig is disabled)? 

> This is done only once, when the ensemble is initial.

Isn't the config version always 0 when dynamic reconfig is disabled?

Reply:
I think I'll revert this part of the change in the next commit.